File: src/bumperbot_py_examples/bumperbot_py_examples/simple_service_client.py
import rclpy
from rclpy.node import Node
from bumperbot_msgs.srv import AddTwoInts
import sys

class SimpleServiceClient(Node):
    def __init__(self, a, b):
        super().__init__('simple_service_client')
        self.client_ = self.create_client(AddTwoInts, 'add_two_ints')

        while not self.client_.wait_for_service(1.0):
            self.get_logger().info('service not available, waiting again...')

        self.req = AddTwoInts.Request()
        self.req.a = a
        self.req.b = b

        self.future_ = self.client_.call_async(self.req)
        self.future_.add_done_callback(self.future_callback)

    def future_callback(self, future):
        self.get_logger().info('Result of add_two_ints: %d' % future.result().sum)


def main():
    rclpy.init()

    if len(sys.argv) < 3:
        print('Usage: simple_service_client.py <a> <b>')
        return -1
    
    node = SimpleServiceClient(int(sys.argv[1]), int(sys.argv[2]))
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/bumperbot_cpp_examples/src/simple_service_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <bumperbot_msgs/srv/add_two_ints.hpp>

using namespace std::chrono_literals;
using std::placeholders::_1;

class SimpleServiceClient : public rclcpp::Node
{
public:
    SimpleServiceClient(int a, int b) : Node("simple_service_client")
    {
        client_ = create_client<bumperbot_msgs::srv::AddTwoInts>("add_two_ints");

        auto request = std::make_shared<bumperbot_msgs::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        while (!client_->wait_for_service(1s))
        {
            if (!rclcpp::ok())
            {
                RCLCPP_ERROR(get_logger(), "Interrupted while waiting for the service. Exiting.");
                return;
            }
            RCLCPP_INFO(get_logger(), "service not available, waiting again...");
        }

        auto result = client_->async_send_request(request, std::bind(&SimpleServiceClient::responseCallback, this, _1));
    }

private:
    rclcpp::Client<bumperbot_msgs::srv::AddTwoInts>::SharedPtr client_;

    void responseCallback(rclcpp::Client<bumperbot_msgs::srv::AddTwoInts>::SharedFuture future)
    {
        if (future.valid())
        {
            RCLCPP_INFO_STREAM(get_logger(), "Service Response: " << future.get()->sum);
        }

        else
        {
            RCLCPP_ERROR(get_logger(), "Failed to get response from service.");
        }
    }
};


int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    if (argc != 3)
    {
        RCLCPP_ERROR(rclcpp::get_logger("simple_service_client"), "Usage: simple_service_client a b");
        return 1;
    }
    auto node = std::make_shared<SimpleServiceClient>(atoi(argv[1]), atoi(argv[2]));
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}